src/services/keystore.ts

```diff
@@ -155,61 +172,162 @@ export class TrustedKeyStoreService {
+                    break;
+                case threema.PushTokenPrefix.Apns:
+                    type = threema.PushTokenType.Apns;
+                    break;
+                case threema.PushTokenPrefix.Hms:
+                    type = threema.PushTokenType.Hms;
+                    break;
+                default:
+                    this.log.error('Invalid push token type:', encodedString[0]);
+                    return null;
+            }
+            token = encodedString.slice(2);
+        } else if (encodedString !== null) {
+            // Compat
+            token = encodedString;
+            type = threema.PushTokenType.Fcm;
+        }
+        return {token: token, type: type};
+    }
+
+    private async migrateDeprecatedStorageFormat(parts: [string, string], password: string): Promise<void> {
+        const nonce = hexToU8a(parts[0]);
+        const encrypted = hexToU8a(parts[1]);
+        const decrypted = nacl.secretbox.open(
+            encrypted,
+            nonce,
+            this.deprecatedPwToKey((new TextEncoder()).encode(password)),
+        );
+        if (!decrypted) {
+            throw new Error('Unable to decrypt deprecated storage');
+        }
+        const token = this.decodePushToken((decrypted as Uint8Array).slice(96));
+        const ownSecretKey = (decrypted as Uint8Array).slice(32, 64);
+        const peerPublicKey = (decrypted as Uint8Array).slice(64, 96);
+        this.log.warn('Migrating deprecated key storage to scrypt-based key storage');
+        await this.storeTrustedKey(ownSecretKey, peerPublicKey, token.token, token.type, password, this.isAutoSession());
+    }
+
     /**
      * Retrieve the trusted key from local browser storage. Decrypt it using
      * the provided password.
      */
-    public retrieveTrustedKey(password: string): threema.TrustedKeyStoreData | null {
+    public async retrieveTrustedKey(password: string): Promise<threema.TrustedKeyStoreData | null> {
         const storedValue: string = this.storage.getItem(this.STORAGE_KEY);
         if (storedValue === null) {
             return null;
         }
 
+        // Check if the deprecated format is being used. If so, migrate it immediately.
         const parts: string[] = storedValue.split(':');
-        if (parts.length !== 2) {
-            return null;
+        if (parts.length === 2) {
+            try {
+                await this.migrateDeprecatedStorageFormat(parts as [string, string], password);
+            } catch (error) {
+                this.log.error(error);
+                return null;
+            }
         }
 
-        const nonce = hexToU8a(parts[0]);
-        const encrypted = hexToU8a(parts[1]);
-        const decrypted = nacl.secretbox.open(encrypted, nonce, this.pwToKey(password));
-        if (!decrypted) {
+        // Parse unencrypted storage data
+        let raw: any;
+        try {
+            raw = JSON.parse(this.storage.getItem(this.STORAGE_KEY));
```

Comment:
Here we're re-reading the data from storage even if no migration took place.

Instead, you could re-read only if the migration happened, and assign the value to `storedValue`.